modules/core/04-channel/keeper/upgrade.go

```diff
@@ -95,6 +96,188 @@ func (k Keeper) WriteUpgradeInitChannel(
+	if !found {
+		return 0, "", errorsmod.Wrapf(types.ErrChannelNotFound, "port ID (%s) channel ID (%s)", portID, channelID)
+	}
+
+	if !collections.Contains(channel.State, []types.State{types.OPEN, types.INITUPGRADE}) {
+		return 0, "", errorsmod.Wrapf(types.ErrInvalidChannelState, "expected one of [%s, %s], got %s", types.OPEN, types.INITUPGRADE, channel.State)
+	}
+
+	if !k.scopedKeeper.AuthenticateCapability(ctx, chanCap, host.ChannelCapabilityPath(portID, channelID)) {
+		return 0, "", errorsmod.Wrapf(types.ErrChannelCapabilityNotFound, "caller does not own capability for channel, port ID (%s) channel ID (%s)", portID, channelID)
+	}
+
+	if proposedUpgradeChannel.Counterparty.PortId != channel.Counterparty.PortId ||
+		proposedUpgradeChannel.Counterparty.ChannelId != channel.Counterparty.ChannelId {
+		return 0, "", errorsmod.Wrap(types.ErrInvalidChannel, "proposed channel upgrade is invalid")
+	}
+
+	if !channel.Ordering.SubsetOf(proposedUpgradeChannel.Ordering) {
+		return 0, "", errorsmod.Wrap(types.ErrInvalidChannelOrdering, "channel ordering must be a subset of the new ordering")
+	}
+
+	if counterpartyChannel.Ordering != proposedUpgradeChannel.Ordering {
+		return 0, "", errorsmod.Wrapf(types.ErrInvalidChannelOrdering, "channel ordering of counterparty channel and proposed channel must be equal")
+	}
+
+	connectionEnd, err := k.GetConnection(ctx, proposedUpgradeChannel.ConnectionHops[0])
+	if err != nil {
+		return 0, "", err
+	}
+
+	if connectionEnd.GetState() != int32(connectiontypes.OPEN) {
+		return 0, "", errorsmod.Wrapf(
+			connectiontypes.ErrInvalidConnectionState,
+			"connection state is not OPEN (got %s)", connectiontypes.State(connectionEnd.GetState()).String(),
+		)
+	}
+
+	if connectionEnd.GetCounterparty().GetConnectionID() != counterpartyChannel.ConnectionHops[0] {
+		return 0, "", errorsmod.Wrapf(connectiontypes.ErrInvalidConnection, "unexpected counterparty channel connection hops, expected %s but got %s", connectionEnd.GetCounterparty().GetConnectionID(), counterpartyChannel.ConnectionHops[0])
+	}
+
+	if err := k.connectionKeeper.VerifyChannelState(ctx, connectionEnd, proofHeight, proofChannel, proposedUpgradeChannel.Counterparty.PortId,
+		proposedUpgradeChannel.Counterparty.ChannelId, counterpartyChannel); err != nil {
+		return 0, "", err
+	}
+
+	upgradeTimeout := types.UpgradeTimeout{TimeoutHeight: timeoutHeight, TimeoutTimestamp: timeoutTimestamp}
+	if err := k.connectionKeeper.VerifyChannelUpgradeTimeout(ctx, connectionEnd, proofHeight, proofUpgradeTimeout, proposedUpgradeChannel.Counterparty.PortId,
+		proposedUpgradeChannel.Counterparty.ChannelId, upgradeTimeout); err != nil {
+		return 0, "", err
+	}
+
+	if err := k.connectionKeeper.VerifyChannelUpgradeSequence(ctx, connectionEnd, proofHeight, proofUpgradeSequence, proposedUpgradeChannel.Counterparty.PortId,
+		proposedUpgradeChannel.Counterparty.ChannelId, counterpartyUpgradeSequence); err != nil {
+		return 0, "", err
+	}
```

Comment:
We should be doing state verification of counterparty using the original channel connection. This is currently using the connection looked up from the proposed upgrade channel connection hops.

This might look a little ugly as depending on whether or not we are in a crossing hello then we'll need to retrieve the connection from the either current channel (state: OPEN) or the restore channel (crossing hello, state: INITUPGRADE).

ref: https://github.com/cosmos/ibc/pull/950#discussion_r1153264262 

cc. @AdityaSripal

Reply:
this is now being handled in the above `getUpgradeTryConnectionEnd` function which handles the crossing hellos case.